### DISCUSSION TITLE
extra_tags for Discogs-plugin, and extended autotagger candidate information

### DISCUSSION THREAD

**Discussion by sindreruud** [beetbox/beets#4926] [Q&A] [answered]
Hi!

So this is a two-parter.

1.
Musicbrainz got a new configuration option fairly recently, where you can specify `extra_tags` that it will use to match your music. For me this is really helpful because I often have to manually find the realease with the correct `catalognum` and `media`. So after adding: 
```
extra_tags: [media, year, label, catalognum, country] 
```
I've seen a huge improvement in accuracy on the few albums I have imported so far. 

Unfortunately for me, well over 50% of my albums are not on musicbrainz, but 99% of the time they are on Discogs. It would be amazing to have this option in the Discogs-plugin aswell. Took a quick look in `discogs.py` and found this
```
def candidates(self, items, artist, album, va_likely, extra_tags=None):
```
Would this be a good place to start if I want to look into this myself? 

2.
The output from the autotagger is sometimes a bit ambiguous when several discogs release look more or less the same (white label vs test pressing vs actual release for instance). Here is in example that actually illustrates the thing I want to solve in both 1. and 2. The autotagger gives equal match to two bandcamp results, aswell as two discogs results.

```
/incoming/music-beets/[12ROB49] Red Seal - Spinning Top [FLAC-VINYL] [1996] (3 items)
originquery: Using origin file /incoming/music-beets/[12ROB49] Red Seal - Spinning Top [FLAC-VINYL] [1996]/origin.yaml
originquery: ╔════════════════╤═════════════╤══════════════╗
originquery: ║ Field          │ Tagged Data │ Origin Data  ║
originquery: ╟────────────────┼─────────────┼──────────────╢
originquery: ║ Media          │             │ Vinyl        ║
originquery: ║ Edition year   │ 1996        │ 1996         ║
originquery: ║ Record label   │             │ Robs Records ║
originquery: ║ Catalog number │             │ 12ROB49      ║
originquery: ╚════════════════╧═════════════╧══════════════╝
Correcting tags from:
    Red Seal - Spinning Top
To:
    Red Seal - Spinning Top EP
URL:
    https://mintconditionrecordslondon.bandcamp.com/album/spinning-top-ep
(Similarity: 89.2%) (year, label) (bandcamp, Digital Media, 2023, GB, Mint Condition)
 * Spinning Top (#0) -> Spinning Top (#1)
 * Twister (#0)      -> Twister (#2)
 * Hula Loop (#0)    -> Hula Loop (#3)
[A]pply, More candidates, Skip, Use as-is, as Tracks, Group albums,
Enter search, enter Id, aBort, eDit, edit Candidates? M
Finding tags for album "Red Seal - Spinning Top".
Candidates:
1. Red Seal - Spinning Top EP (89.2%) (year, label) (bandcamp, Digital Media, 2023, GB, Mint Condition)
2. Red Seal - Spinning Top EP (89.2%) (year, label) (bandcamp, Vinyl, 2023, GB, Mint Condition)
3. Red Seal - Spinning Top (86.2%) (tracks, source) (Discogs, Vinyl, 1996, UK, Robs Records, 12ROB49)
4. Red Seal - Spinning Top (86.2%) (tracks, source) (Discogs, Vinyl, 1996, UK, Robs Records, 12ROB49)
5. Red Seal - Spinning Top  (79.3%) (tracks, source, catalognum, ...) (Discogs, Vinyl, 2023, UK, Mint Condition (2), MC061)
6. Spinning Top - Spinning Top (37.5%) (tracks, artist, missing tracks, ...) (bandcamp, Digital Media, 2017, US, Light Gray Records)
7. Boston Symphony Orchestra ... Charles Munch - The French Touch (23.8%) (tracks, missing tracks, artist, ...) (Discogs, Vinyl, 1959, US, RCA Victor Red Seal, LSC-2292)
8. Michala Petri , The London Philharmonic Orchestra - Scandinavian Moods (18.1%) (missing tracks, tracks, artist, ...) (Discogs, CD, 1999, Europe, RCA Red Seal, 743211679502)
# selection (default 1), Skip, Use as-is, as Tracks, Group albums,
Enter search, enter Id, aBort, eDit, edit Candidates?
```

In this scenario I know that I want to pick one of the Discogs candidates because they in my opinion have better metadata. But candidate 4 and 5 does not appear different.  But the difference is:
Candidate 3: [r74882] (Vinyl, 12", 45 RPM)
Canditate 4: [r4383728] (Vinyl, 12", 45 RPM, White Label)

So i wonder how hard it would be to customize the output of this string
`(Discogs, Vinyl, 1996, UK, Robs Records, 12ROB49)`
so it uses the entire `Format` and looks something like this:
`(Discogs, Vinyl, 12", 45 RPM, White Label, 1996, UK, Robs Records, 12ROB49)` 

Happy to look into this myself, but would be very grateful if someone could nudge me in the right direction as this is a bit outside my comfort-zone.

**Comment by arsaboo** [comment]
Yes, you would start there and modify the get_albums function to use the extra tags. You may want to check whether the underlying python3-discogs-client library can use extra tags. Looks like it can accept a few additional tags 

https://www.discogs.com/developers/#page:database,header:database-search

**Reply by JOJ0** [reply]
Yeah go for it @sindreruud I'll review your PR. Will be a great addition to the Discogs plugin😍

**Reply by JOJ0** [reply]
And yes python3-discogs-client supports searching for everything the Discogs API allows: https://python3-discogs-client.readthedocs.io/en/latest/quickstart.html#searching

**Comment by sindreruud** [comment]
1,5 years and I'm back taking a stab at this 😁  Might not be the best place to ask, but I'll go for it anyway.

I have tested the discogs_client via cli and done queries like this
```
>>> import discogs_client
>>> d = discogs_client.Client('ExampleApplication/0.1', user_token="snip")
>>> results = d.search('Red Seal - Spinning Top', type='release', catno='12ROB49', year='1996')
>>> print(results.page(1))
[<Release 74882 'Red Seal - Spinning Top'>, <Release 4383728 'Red Seal - Spinning Top'>]
```

Looks quite good. But I am sure there are corner cases that needs to be tackled, like spaces, upper/lower case, etc.

Anyway. To get started, much like the autotagger does, i made a dict to map the fields
```
FIELDS_TO_DISCOGS_KEYS = {
    "catalognum": "catno",
    "country": "country",
    "label": "label",
    "barcode": "barcode",
    "media": "format",
    "year": "year",
}
```

And then I took a stab at changing the candidates()
```
def candidates(self, items, artist, album, va_likely, extra_tags: dict[str, Any] | None = None):
        """Returns a list of AlbumInfo objects for discogs search results
        matching an album and artist (if not various).
        """
        if not self.discogs_client:
            return

        if not album and not artist:
            self._log.debug(
                "Skipping Discogs query. Files missing album and "
                "artist tags."
            )
            return []

        if va_likely:
            query = album
        else:
            query = f"{artist} {album}"

        # Additional search cues from existing metadata.
        if extra_tags:
            tags = []
            for tag, value in extra_tags.items():
                key = FIELDS_TO_DISCOGS_KEYS[tag]
                if key and value:
                    tags.append(f"{key}='{value}'")

        # Append extra filters to the query
        if tags:
            query += ", " + ", ".join(tags)
        
        try:
            return self.get_albums(query)
        except DiscogsAPIError as e:
            self._log.debug("API Error: {0} (query: {1})", e, query)
            if e.status_code == 401:
                self.reset_auth()
                return self.candidates(items, artist, album, va_likely)
            else:
                return []
        except CONNECTION_ERRORS:
            self._log.debug("Connection error in album search", exc_info=True)
            return []
```

And now comes the stupid question, so please bear with me. But how do I go about testing this? Tried following the developer docs and the CONTRIBUTING.rst. So I have a virtual python environment with poetry, poethepoet and I did poetry install. But how do i go about actually building/installing (right terminology?) beets with my changes so I can start testing it?

If someone could nudge me in the right direction I would be very happy :)

**Reply by arsaboo** [reply]
I just install beets from the specific branch and test it out. If your branch is `fix`, you can use something like (please update the git URL):

`pip install --upgrade --force-reinstall --no-deps git+https://github.com/arsaboo/beets.git@fix`

**Reply by sindreruud** [reply]
Thanks, real simple solution! :)

**Reply by JOJ0** [reply]
> Thanks, real simple solution! :)

There's an even simpler one. Go to your local repo clone directory and install like this:

```
pip install -e .
```


-e is editable mode. you can immediately run a `beet something` command that includes your uncommited code changes.

**Reply by jdoe29103** [reply]
@sindreruud were you ever able to get this working? I also have the same issue where I have the catalog number for a release which has a 100% match on Discogs but is not found by the auto tagger due the number of release on Discogs being greater than my search_limit.

If there is anything I can do to help contribute and get this merged please let me know!

**Reply by sindreruud** [reply]
I did get started on it: https://github.com/sindreruud/beets/tree/discogs-extra-tags

However, my limited coding knowledge made me rely to much on vibecoding and I sort of lost the plot. I was able to supply the discogs query with the label information, but couldn't quite get the weighting or distance stuff to work properly. As far as I can remember.. Would be keen to look at this again in the future. Unfortunately that won't be soon. Feel free to use what you can if you want to continue.

**Reply by jdoe29103** [reply]
@sindreruud I'll take a look at it! I am really only interested in getting the ability to search with catalognum. Everything else is just a bonus. Can you tell me if you got extra_tags working at all? When you say you couldn't quite get the weighting or distance stuff working I'm curious what you would even need to change or order to get extra_tags working.

**Reply by jdoe29103** [reply, marked as answer]
@sindreruud I think I now have this working. Have a look over my PR if you're interested. #6416

**Reply by sindreruud** [reply]
Very cool! Did it get closed? Any plans to get this merged? I have some time in the coming weekend to test this if that is helpful.

Edit: Nevermind, I see it got merged in #6416 
Thanks a bunch, this is really helpfull for my workflow. Good work!
